### PR TITLE
Change unknown_service to DefaultServiceName

### LINKED ISSUE #5650: Unknown service created when using Hangfire instrumentation
**Describe the bug**
We use the Hangfire OTEL instrumentation ([located here](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Hangfire)) to add root spans to the running jobs.

OTEL is configured like so to add the .NET Activity instrumentation but the export is delegated to DD Trace's automated OTEL support:
```cs
using var tracer = Sdk
    .CreateTracerProviderBuilder()
    .AddHangfireInstrumentation()
    .Build();
```

However this results in the hangfire root spans showing up under their own service "unknown_service:dotnet" in the DD console. All other spans from the service are correctly aggregated.

image

The span metadata does have the correct service name:
image

This application is running in App Services with the tracer baked into the image:

```
COPY --from=datadog/serverless-init:1 /datadog-init /app/datadog-init
COPY --from=datadog/dd-lib-dotnet-init /datadog-init/monitoring-home/ /dd_tracer/dotnet/
ENV DD_SERVICE=(hidden)-integration-api
ENV DD_LOGS_INJECTION=true
ENV DD_LOGS_DIRECT_SUBMISSION_INTEGRATIONS='Serilog'
ENV DD_TRACE_OTEL_ENABLED=true
ENV DD_DATA_STREAMS_ENABLED=true
ENV DD_TRACE_SPAN_ATTRIBUTE_SCHEMA='v1'
```

**Runtime environment:**
- Instrumentation mode: Automatic
- Tracer version: 2.50.0
- OS: Azure App Services (Linux)
- CLR: .NET 8

### PR DESCRIPTION
**Pull request by bouwkast**
## Summary of changes

This looks for an `unknown_service` in `service.name` for `Activity`/`OpenTelemetry` spans and if detected will replace that with the value of `DefaultServiceName`.

## Reason for change

OpenTelemetry will use `unknown_service:AssemblyName` as the default for `service.name` if there isn't a value for it.
A workaround to this is to have `OTEL_SERVICE_NAME` set to the same value as `DD_ENV`, this avoids needing to do that workaround.

## Implementation details

Avoid `unknown_service` https://opentelemetry.io/docs/specs/semconv/resource/#service

## Test coverage

Updated the OpenTelemetrySdk tests to replicate this.

## Other details

Fixes #5650

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5671)](https://github.com/DataDog/dd-trace-dotnet/tree/da5a2099674b9e87fb4b7a7b60fe06c8f835473b)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/90483128dd2be87c7bc028820745f7f36df4f7c3)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (73ms)  : 66, 80
     .   : milestone, 73,
    master - mean (73ms)  : 64, 83
     .   : milestone, 73,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (979ms)  : 956, 1002
     .   : milestone, 979,
    master - mean (981ms)  : 964, 999
     .   : milestone, 981,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (108ms)  : 106, 111
     .   : milestone, 108,
    master - mean (109ms)  : 105, 113
     .   : milestone, 109,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (682ms)  : 660, 704
     .   : milestone, 682,
    master - mean (691ms)  : 668, 715
     .   : milestone, 691,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (92ms)  : 89, 95
     .   : milestone, 92,
    master - mean (94ms)  : 90, 98
     .   : milestone, 94,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (643ms)  : 623, 663
     .   : milestone, 643,
    master - mean (650ms)  : 632, 667
     .   : milestone, 650,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (198ms)  : 194, 202
     .   : milestone, 198,
    master - mean (192ms)  : 188, 197
     .   : milestone, 192,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (1,094ms)  : 1072, 1116
     .   : milestone, 1094,
    master - mean (1,079ms)  : 1052, 1106
     .   : milestone, 1079,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (289ms)  : 282, 296
     .   : milestone, 289,
    master - mean (278ms)  : 272, 284
     .   : milestone, 278,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (892ms)  : 862, 921
     .   : milestone, 892,
    master - mean (860ms)  : 838, 883
     .   : milestone, 860,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5671) - mean (274ms)  : 267, 281
     .   : milestone, 274,
    master - mean (268ms)  : 263, 272
     .   : milestone, 268,

    section CallTarget+Inlining+NGEN
    This PR (5671) - mean (862ms)  : 833, 891
     .   : milestone, 862,
    master - mean (855ms)  : 825, 886
     .   : milestone, 855,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5671)](https://github.com/DataDog/dd-trace-dotnet/tree/da5a2099674b9e87fb4b7a7b60fe06c8f835473b)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/2545a4b37a7b6b99944bf29a1469f925a4249a3c)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5671) (11.751M)   : 0, 11751492
    master (11.733M)   : 0, 11733185
    benchmarks/2.9.0 (11.919M)   : 0, 11919161

    section Automatic
    This PR (5671) (7.998M)   : 0, 7998457
    master (8.040M)   : 0, 8040098
    benchmarks/2.9.0 (8.337M)   : 0, 8336508

    section Trace stats
    master (8.324M)   : 0, 8323692

    section Manual
    This PR (5671) (10.244M)   : 0, 10243980
    master (10.054M)   : 0, 10054033

    section Manual + Automatic
    This PR (5671) (7.529M)   : 0, 7529483
    master (7.463M)   : 0, 7462591

    section Version Conflict
    master (6.715M)   : 0, 6715141

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5671) (9.514M)   : 0, 9513738
    master (9.631M)   : 0, 9630641
    benchmarks/2.9.0 (9.646M)   : 0, 9645709

    section Automatic
    This PR (5671) (6.654M)   : 0, 6654443
    master (6.456M)   : 0, 6456385

    section Trace stats
    master (6.954M)   : 0, 6953975

    section Manual
    This PR (5671) (8.255M)   : 0, 8254610
    master (8.240M)   : 0, 8239810

    section Manual + Automatic
    This PR (5671) (6.154M)   : 0, 6154011
    master (5.934M)   : 0, 5934116

    section Version Conflict
    master (5.697M)   : 0, 5697158

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5671) (9.919M)   : 0, 9918792
    benchmarks/2.9.0 (10.036M)   : 0, 10035695

    section Automatic
    This PR (5671) (7.081M)   : 0, 7081444
    benchmarks/2.9.0 (7.475M)   : 0, 7475419

    section Manual
    This PR (5671) (8.849M)   : 0, 8848960

    section Manual + Automatic
    This PR (5671) (6.866M)   : 0, 6865732

```

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/OpenTelemetry/ResourceAttributeProcessorHelper.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using System.Reflection.Emit;
 using Datadog.Trace.Activity.DuckTypes;
 using Datadog.Trace.Activity.Handlers;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.DynamoDb;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/OpenTelemetry/ResourceAttributeProcessorHelper.cs:12]: Looks like an accident to me 🙂 
```suggestion
```

File: tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/OpenTelemetry/ResourceAttributeProcessorHelper.cs [hunk 2]
```diff
@@ -58,7 +59,16 @@ public static void OnStart(object processor, object activityData)
                             {
                                 if (attribute.Key == "service.name")
                                 {
-                                    span.ServiceName = attribute.Value.ToString();
+                                    var resourceServiceName = attribute.Value.ToString();
+
+                                    if (string.IsNullOrEmpty(resourceServiceName) || resourceServiceName.Contains("unknown_service"))
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/OpenTelemetry/ResourceAttributeProcessorHelper.cs:64]: nit: 
```suggestion
                                    if (string.IsNullOrEmpty(resourceServiceName)
                                     || string.Equals(resourceServiceName, "unknown_service", StringComparison.Ordinal)
                                     || resourceServiceName.StartsWith("unknown_service:", StringComparison.Ordinal))
```

File: tracer/test/snapshots/OpenTelemetrySdkTests.verified.txt [hunk 3]
```diff
@@ -949,6 +949,33 @@ at Samples.OpenTelemetrySdk.Program.RunSpanUpdateMethods(TelemetrySpan span),
     TraceId: Id_40,
     SpanId: Id_41,
     Name: internal,
+    Resource: service.name should be the DefaultServiceName value,
```
Comment (andrewlock) [inline tracer/test/snapshots/OpenTelemetrySdkTests.verified.txt:952]: I  really like this approach of having the verbose "assertion" in the Resource to make it easier to check the spans 👍

**Review by andrewlock — APPROVED** [review]
:shipit:

**Review by lucaspimentel — APPROVED** [review]

**Pull request merged**